src/patterns/kinesis-lambda.ts

```diff
@@ -0,0 +1,111 @@
+import type { StreamProps } from "@aws-cdk/aws-kinesis";
+import { Stream } from "@aws-cdk/aws-kinesis";
+import type { IEventSourceDlq } from "@aws-cdk/aws-lambda";
+import { StartingPosition } from "@aws-cdk/aws-lambda";
+import type { KinesisEventSourceProps } from "@aws-cdk/aws-lambda-event-sources";
+import { KinesisEventSource } from "@aws-cdk/aws-lambda-event-sources";
+import { Duration } from "@aws-cdk/core";
+import type { GuLambdaErrorPercentageMonitoringProps, NoMonitoring } from "../constructs/cloudwatch";
+import type { GuStack } from "../constructs/core";
+import type { GuKinesisStreamProps } from "../constructs/kinesis";
+import { GuKinesisStream } from "../constructs/kinesis";
+import type { GuFunctionProps } from "../constructs/lambda";
+import { GuLambdaFunction } from "../constructs/lambda";
+
+export interface ExistingKinesisStream {
+  logicalIdFromCloudFormation?: string;
+  externalKinesisStreamName?: string;
+}
+
+export type StreamProcessingProps = Omit<
+  KinesisEventSourceProps,
+  "bisectBatchOnError" | "maxRecordAge" | "onFailure" | "retryAttempts"
+>;
+
+type AwsErrorHandlingProps = Pick<
+  KinesisEventSourceProps,
+  "bisectBatchOnError" | "maxRecordAge" | "onFailure" | "retryAttempts"
+>;
+
+function toAwsErrorHandlingProps(errorHandlingProps: ErrorHandlingProps): AwsErrorHandlingProps {
+  return {
+    bisectBatchOnError: errorHandlingProps.bisectBatchOnError,
+    onFailure: errorHandlingProps.deadLetterQueueForSkippedRecords,
+    ...errorHandlingProps.retryBehaviour.toAwsProp(),
+  };
+}
+
+type AwsRetryProp = Pick<AwsErrorHandlingProps, "maxRecordAge" | "retryAttempts">;
+type RetryType = "attempts" | "recordAge";
+
+export class Retry {
```

Comment:
I think it'd be nice for patterns to be super simple to make it easy to follow, both in terms of their orchestration of constructs but also in terms of their source code - can we move this class into it's own file? I think some of the interfaces could also be moved too.

Reply:
Good idea. I've moved a lot of the 'helper' code into [its own file](https://github.com/guardian/cdk/pull/296/files#diff-c4842322ffd7111821d75fbe3cce9aecb7f630bc22e877934ae644a947f3c9d1) at the construct level.